pwnlib/windbg.py

```diff
@@ -0,0 +1,239 @@
+        exe(str): Path to the executable on disk.
+        env(dict): Environment to start the binary in.
+        creationflags(int): Flags to pass to :func:`.process.process`.
+
+    Returns:
+        :class:`.process`: A tube connected to the target process.
+
+    Notes:
+
+        .. code-block: python
+
+            # Create a new process, and stop it at 'main'
+            io = windbg.debug('calc', '''
+            bp $exentry
+            go
+            ''')
+
+        When WinDbg opens via :func:`.debug`, it will initially be stopped on the very first
+        instruction of the entry point.
+    """
+    if isinstance(
+        args, six.integer_types + (tubes.process.process, tubes.ssh.ssh_channel)
+    ):
+        log.error("Use windbg.attach() to debug a running process")
+
+    if context.noptrace:
+        log.warn_once("Skipping debugger since context.noptrace==True")
+        return tubes.process.process(args, executable=exe, env=env, creationflags=creationflags)
+    
+    windbgscript = windbgscript or ''
+    if isinstance(windbgscript, six.string_types):
+        windbgscript = windbgscript.split('\n')
+    # resume main thread
+    windbgscript = ['~0m'] + windbgscript
+    creationflags |= CREATE_SUSPENDED
+    io = tubes.process.process(args, executable=exe, env=env, creationflags=creationflags)
+    attach(target=io, windbgscript=windbgscript, **kwargs)
+
+    return io
+
+def binary():
+    """binary() -> str
+
+    Returns the path to the WinDbg binary.
+
+    Returns:
+        str: Path to the appropriate ``windbg`` binary to use.
+    """
+    windbg = misc.which('windbgx.exe') or misc.which('windbg.exe')
+    if not windbg:
+        log.error('windbg is not installed or in system PATH')
+    return windbg
+
+@LocalContext
+def attach(target, windbgscript=None, windbg_args=[]):
+    """attach(target, windbgscript=None, windbg_args=[]) -> int
+
+    Attach to a running process with WinDbg.
+
+    Arguments:
+        target(int, str, process): Process to attach to.
+        windbgscript(str, list): WinDbg script to run after attaching.
+        windbg_args(list): Additional arguments to pass to WinDbg.
+
+    Returns:
+        int: PID of the WinDbg process.
+
+    Notes:
+
+        The ``target`` argument is very robust, and can be any of the following:
+
+        :obj:`int`
+            PID of a process
+        :obj:`str`
+            Process name.  The youngest process is selected.
+        :class:`.process`
+            Process to connect to
+    
+    Examples:
+
+        Attach to a process by PID
+
+        >>> pid = windbg.attach(1234) # doctest: +SKIP
+
+        Attach to the youngest process by name
+
+        >>> pid = windbg.attach('cmd.exe') # doctest: +SKIP
+
+        Attach a debugger to a :class:`.process` tube and automate interaction
+
+        >>> io = process('cmd') # doctest: +SKIP
+        >>> pid = windbg.attach(io, windbgscript='''
+        ... bp ucrtbase:puts

```

Comment:
Typo in command syntax should be ``ucrt!puts``.
A better example would be kernelbase!WriteFile since cmd.exe would write using WriteFile if stdout is not a console.